### PR TITLE
[SPARK-36498][SQL] Reorder inner fields of the input query in byName V2 write

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

Today, when we write data to a v2 table with byName mode, we only reorder the top-level columns, not inner struct fields. This doesn't make sense as Spark should treat inner struct fields as the first-class citizen (e.g. nested column pruning, filter pushdown with nested columns).

This PR improves `TableOutputResolver` to reorder inner fields as well.

### Why are the changes needed?

better user-experience

### Does this PR introduce _any_ user-facing change?

yes, more queries are allowed to write to v2 tables.

### How was this patch tested?

new test

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @sunchao @yaooqinn

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala [hunk 1]
```diff
@@ -74,6 +66,42 @@ object TableOutputResolver {
...
+        }
+        (matchedCol.dataType, expectedCol.dataType) match {
+          case (input: StructType, expected: StructType) =>
+            val fields = input.zipWithIndex.map { case (f, i) =>
+              Alias(GetStructField(matchedCol, i, Some(f.name)), f.name)()
+            }
+            val reordered = reorderColumnsByName(
+              fields, expected.toAttributes, conf, addError, newColPath)
+            Some(Alias(CreateStruct(reordered), expectedCol.name)())
+          case _ =>
```
Comment (dilipbiswal) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:98]: @cloud-fan Thanks for the fix !! I am thinking, do we need cases for Array and Map types ? Like for cases.
Struct<a: Array<Struct>> . Don't we need to create the arrays and maps from the remapped inner structs ?

Comment (sunchao) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:98]: I think these case are already covered in the clause.

**Review by sunchao — APPROVED** [review]
Thanks @cloud-fan for pinging. LGTM (non-binding).

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by yaooqinn** [comment]
retest this please

**Review by yaooqinn — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:98]: Let me add tests and check it out.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:107]: Added

Comment (sunchao) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:107]: oops I missed the cases when structs are inside array & map, good point @dilipbiswal !

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala [hunk 2]
```diff
@@ -74,6 +66,160 @@ object TableOutputResolver {
...
+        AttributeReference("v", expectedType.valueType, expectedType.valueContainsNull)()
+      val resValue = reorderColumnsByName(
+        Seq(valueParam), Seq(fakeValueAttr), conf, addError, colPath :+ "value")
+
+      if (resKey.length == 1 && resValue.length == 1) {
+        val keyFunc = LambdaFunction(resKey.head, Seq(keyParam))
+        val valueFunc = LambdaFunction(resValue.head, Seq(valueParam))
+        val newKeys = ArrayTransform(MapKeys(input), keyFunc)
+        val newValues = ArrayTransform(MapValues(input), valueFunc)
+        Some(Alias(MapFromArrays(newKeys, newValues), expectedName)())
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:216]: `TransformValues(TransformKeys(input, keyFunc), valueFunc)`?

**Review by viirya — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:216]: This creates map twice, can be slower.

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:216]: ok

**Comment by cloud-fan** [comment]
jenkins passes and GA failure is unrelated. I'm merging this to master, thanks for the review!

**Pull request closed without merging**